### DISCUSSION TITLE
NavigationStackStore missunderstanding

### DISCUSSION THREAD

**Discussion by ivangodfather** [pointfreeco/swift-composable-architecture#2131] [Q&A] [answered]
I have a top level `RootFeature` with a TabBar.
In first tab I have `NavigationFeature`, that pushes `ItemFeature` using a regular `NavigationStack` and 
```
            .navigationDestination(
                store: store.scope(state: \.$destination, action: NavigationFeature.Action.destination),
                state: /NavigationFeature.Destination.State.itemAsNavigation,
                action: NavigationFeature.Destination.Action.itemAsNavigation,
                destination: ItemView.init
            )
```

In a second tab I have in the middle called `StackNavFeature`. This feature holds a `NavigationStackStore`, and I'm able to present the `NavigaitonFeature` from first tab. The problem is that now when I try to use the link with the above pattern the link gets stuck and freezes.
I'm not even sure if mixing like this is possible, or should I always use `NavigationStackStore` in both cases?

Here is a sample app showcasing the problem:
https://github.com/ivangodfather/TCANavBug

**Comment by mbrandonw** [comment]
Hi @ivangodfather, I can't say I know exactly why the app is freezing up, but I would say in general you should not mix the `NavigationStackStore` style of navigation with the `presentationDestination(store:)` style. The `NavigationExampleView` is trying to be used in both situations, and that is causing the problem.

Similar things happen when trying to mix `NavigationStack` with `navigationDestination(isPresented:)` in vanilla SwiftUI. It doesn't freeze, but it is possible for SwiftUI to become confused as to which screen is actually being presented since the one shown with `navigationDestination` is not represented in the stack's binding, and it can leave the UI in an inconsistent state.

Also, I am a little confused with the comment in your code:

```swift
// I image the bug is here, when I click the item feature (embeeded inside navigation) it goes the with navigation switch
```

The `NavigationLink(state:)` you are using is pointed to the `navigation` case:

```swift
NavigationLink(state: StackNavFeature.Path.State.navigation(.init()))
```

That is why it is destructuring on the `navigation` case.

**Reply by ivangodfather** [reply]
@mbrandonw this comment is because when I tap the **item as a link** the `NavigationStackStore` destination also gets called, but not entering the case `.item`, instead the case `.navigation`. 

From your answer I understand the solution is to use a `NavigationStackStore` in both cases, but still this feature will live inside 2 differents ones,  

```
        NavigationLink(
          state: StackNavFeature.Path.State.item(
           ....
          )
        ) {
          Text("Go to item")
        }
        NavigationLink(
          state: ???.Path.State.item(
           ....
          )
        ) {
          Text("Go to item")
        }
```

I'm still a bit confused :(

**Reply by mbrandonw** [reply, marked as answer]
Yeah, in that case, when you need a child feature to live in two different stacks, you cannot use `NavigationLink(state:)`. This is true of vanilla SwiftUI too when using an array for the binding in `NavigationStack`.

Instead, you need to send an action from the child feature that the parent domain intercepts, and then the parent reducer will append to the `path` stack state.

**Reply by ivangodfather** [reply]
Oh ok! Thanks! I wasn't aware of this

**Reply by atsixian** [reply]
> in general you should not mix the NavigationStackStore style of navigation with the presentationDestination(store:) style

Hi @mbrandonw, I'm wondering if we can put this into the doc. My program freezes today because of mixing these two styles, but I didn't find any warning in the [Navigation](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/navigation) section. I think it may help others who are not familiar with the vanilla SwiftUI issue like me.
